### ISSUE TITLE
[iOS] Setting BitmapImage Source after it has been assigned to an image does not work

### ISSUE THREAD

**Issue by mikernet** [unoplatform/uno#5453]
## Current behavior

Consider the following code:

```cs
        public BitmapImage GetBitmap()
        {
            var image = new BitmapImage();
            LoadImage();
            return image;

            async void LoadImage()
            {
                await Task.Delay(100);
                image.UriSource = new Uri("ms-appx:///Assets/SomeImage.png");
            }
        }
```

And a XAML binding such as:

```xml
<Image Source="{x:Bind GetBitmap()}"/>
```

The image never gets set. This works fine in UWP. In our actual use-case `GetBitmap()` has an image ID parameter and `LoadImage()` loads the image asynchronously from a web service and uses `image.SetSourceAsync()` instead of `UriSource` but the same problem is exhibited with this simpler demonstration. 

## Expected behavior

The image should show whatever the source is set to later.

## Workaround

Bind to an image source that fires a property changed notification when it changes which then provides a "ready" `BitmapImage`, but this requires considerable changes to the application architecture.

## Environment

Nuget Package:

- [X] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s): 3.5.0-dev.385, 3.6.0-dev.378

Affected platform(s):

- [X] iOS
- [ ] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

## Anything else we need to know?

I'm guessing this may be a problem on other platforms (i.e. Android) as well, but I'm not sure as of right now.

**Comment by jeromelaban** [comment]
Thanks for the report. Can you create a repro sample of the problem ? The code seems simple enough, but it'll be easier to make sure this is actually related to BitmapImage.

**Comment by mikernet** [comment]
Give me a sec with this, I'm trying a few things out to trace the issue - the above sample code actually works, the file path is just case sensitive on iOS (which makes sense) so my sample code above doesn't actually reproduce the issue. It works fine on WASM and UWP but iOS refuses to show the image for some odd reason. I'll follow up on this and edit the problem once I figure it out.

**Comment by mikernet** [comment]
Okay, so setting `SourceUri` actually works fine but `SetSource()` / `SetSourceAsync()` does not:
```cs
public static BitmapImage GetBitmap(ImageType imageType, long? id)
        {
            var image = new BitmapImage();
            LoadImage();
            return image;

            async void LoadImage()
            {
                try {
                    await Task.Delay(100);
                    string fname = @"Assets/Logo.png";
                    StorageFolder InstallationFolder = Windows.ApplicationModel.Package.Current.InstalledLocation;
                    StorageFile file = await InstallationFolder.GetFileAsync(fname);
                    var imageStream = new MemoryStream();
                    using (var fs = File.OpenRead(file.Path)) {
                        fs.CopyTo(imageStream);
                    }

                    // option 1:
                    image.UriSource = new Uri("ms-appx:///Assets/Logo.png");
                    // option 2:
                    image.SetSource(imageStream.AsRandomAccessStream());
                }
                catch (Exception ex) {
                    Console.WriteLine(ex.Message);
                }

                Console.WriteLine("Image set");
            }
        }
```

If I leave all the code untouched except for picking either option 1 or option 2, it works in the former case but fails to show the image in the latter case. It does not hit the exception block in either.

I'll throw a repro project up on github if you like.

**Comment by mikernet** [comment]
I changed the code in my post above a bit - if I comment out the `await Task.Delay` line then it works, so it's definitely an issue with calling `SetSource()` after a delay.

**Comment by mikernet** [comment]
Here's the minimal repro you can just paste into any test project - just change the file name to an image in the assets folder and set an image's source to the return value:

```cs
        public static BitmapImage GetBitmap()
        {
            var image = new BitmapImage();
            LoadImage();
            return image;

            async void LoadImage()
            {
                //await Task.Delay(100);
                string fname = @"Assets/Logo.png";
                StorageFolder InstallationFolder = Windows.ApplicationModel.Package.Current.InstalledLocation;
                using (var fs = File.OpenRead(Path.Combine(InstallationFolder.Path, fname))) {
                    image.SetSource(fs.AsRandomAccessStream());
                }
            }
        }
```

It works until you uncomment the `await` line.

**Comment by dr1rrb** [comment]
It looks like the image is opened only once per consumer. We should multi-target the pattern used for wasm and skia (i.e. the [`Subscribe`](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Media/ImageSource.netstd.cs#L45)), and enable the [`ForceLoad`](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Media/Imaging/BitmapSource.cs#L113) which invokes the [`InvalidateSource()`](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Media/ImageSource.netstd.cs#L81) method which will updates all listeners (i.e. images and brushes).

**Comment by mikernet** [comment]
Could you point me towards the area in source where these changes need to be made?

Unfortunately, there seem to be some layout regressions in 3.6.0-dev.latest that prevent me from being able to use it in iOS so I'm probably going to have to backport the changes to 3.5.1 or figure out what is going on there. I'll try to PR the fix since working around this issue is going to be a royal PITA for how we have things setup across the application.

**Comment by iury-kc** [comment]
@dr1rrb Hi!
I tried to reproduce your code. However, i had to use SetSourceAsync instead SetSource.
Did you tried this one?

**Comment by iury-kc** [comment]
i created a button to set a component Image source like that:

 imgLogo.Source = GetBitmap();

What i could see, there`s a intermitent behavior. When you click a few times, eventually you get a blank image. 
Not sure if this is only in debugging. Still verifying.

**Comment by iury-kc** [comment]
@dr1rrb Hi! 
I tried to reproduce your code. However, SetSource() wont work in any platform, not only iOS.
Using a Static GetImage method and XAML binding didn`t worked as well.   

Using SetSourceAsync() works fine, besides having the intermitence i mentioned las comment. 
Do you mind to confirm your code please? Maybe it has been solved for SetSourceAsync( ) but not yet for SetSource( ).

**Comment by dr1rrb** [comment]
Hi @iury-kc, to which code do you refer? The repro from @mikernet ? 

I don't remember having investigated nor tested this issue, the only thing is that if we are about to significantly change `ImageSource`, I want to make sure we are following the new pattern of the `Subscribe`. A long time ago when we started to dev uno we took a wrong approach to switch on each known kind of `ImageSource` directly in `Image` and `ImageBrush` (to mention only those), which drives us to copy the same code again and again on each platform and having some weird specific issues like this one.

**Comment by iury-kc** [comment]
@dr1rrb Oops. Tagged the wrong. i mean @mikernet

**Comment by iury-kc** [comment]
@mikernet Hi! Do you mind to give more details about what is preventing to use the Subscribe pattern on iOS ? Thanks!

**Comment by iury-kc** [comment]
Hi!
Moving the delegate `internal event EventHandler Invalidated;` from WritableBitmap to ImageSource and calling the OnInvalidated when BitmapImage has changed ( URI, stream ....) solved the problem on iOS.  That make the code cleaner and easier to understand. 
Test is done.
 
The pattern used by SKIA files still doesn`t work. Still investigating what is the problem there. 
As it will take a time to solve for all platforms, should I continue investigation or should I pause?

**Comment by iury-kc** [comment]
Did a PR that should close this issue. 
Also, we have Android similar behavior been solved on issue: https://github.com/unoplatform/uno/issues/7288.
For that we probably will solve for SKIA/WASM as well. If not, will create a separated issue to manage those platforms.

**Comment by Youssef1313** [comment]
https://github.com/unoplatform/uno/pull/12076 might fix this.

**Comment by MartinZikmund** [comment]
No longer relevant, works correctly for Skia renderer targets
